src/imitation/data/types.py

```diff
@@ -22,6 +27,177 @@
+
+    @classmethod
+    def from_obs_list(cls, obs_list: List[Dict[str, np.ndarray]]):
+        return cls(
+            {k: np.stack([obs[k] for obs in obs_list]) for k in obs_list[0].keys()},
+        )
+
+    def __post_init__(self):
+        if not all((isinstance(v, np.ndarray) for v in self.d.values())):
+            raise ValueError("keys must by numpy arrays")
+
+    def __len__(self):
+        """Returns the first dimension of constiuent arrays.
+
+        Only defined if there is at least one array, and all arrays have the same
+        length of first dimension. Otherwise raises ValueError.
+
+        Len of a DictObs usually represents number of timesteps, or number of
+        environments in a VecEnv.
+
+        Use `dict_len` to get the number of entries in the dictionary.
+
+        Raises:
+            ValueError: if the arrays have different lengths or there are no arrays.
+
+        Returns:
+            The length (first dimension) of the constiuent arrays
+        """
+        lens = set(len(v) for v in self.d.values())
+        if len(lens) == 1:
+            return lens.pop()
+        elif len(lens) == 0:
+            raise ValueError("Length not defined as DictObs is empty")
+        else:
+            raise ValueError(
+                f"Length not defined; arrays have conflicting first dimensions: {lens}",
+            )
+
+    @property
+    def dict_len(self):
+        return len(self.d)
+
+    def __getitem__(self, key: Union[slice, int]) -> "DictObs":
+        """Indexes or slices into the first element of every array.
+
+        Note that it will still return singleton values as np.arrays, not scalars,
+        to be consistent with DictObs type signature.
+        Also note that we don't support multi-dimensional slicing.
+
+        Args:
+            key: a single slice
+
+        Returns:
+            A new DictObj object with each array indexed.
+        """
+        # asarray handles case where we slice to a single array element.
+        return self.__class__({k: np.asarray(v[key]) for k, v in self.d.items()})
+
+    def __iter__(self) -> Iterator["DictObs"]:
+        """Iterates over the first dimension of each array."""
+        return (self[i] for i in range(len(self)))
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        if not self.d.keys() == other.d.keys():
+            return False
+        return all(np.array_equal(self.d[k], other.d[k]) for k in self.d.keys())
+
+    @property
+    def shape(self) -> Dict[str, Tuple[int, ...]]:
+        """Returns a dictionary with shape-tuples in place of the arrays."""
+        return {k: v.shape for k, v in self.d.items()}
+
+    @property
+    def dtype(self) -> Dict[str, np.dtype]:
+        """Returns a dictionary with shape-tuples in place of the arrays."""
+        return {k: v.dtype for k, v in self.d.items()}
+
+    def unwrap(self) -> Dict[str, np.ndarray]:
+        return self.d
```

Comment:
Do we want to allow callers to mutate our internal dict? (This is a genuine question, I'm not sure.)

Reply:
Now returns a copy of the dict. It doesn't copy arrays, so still possible to mutate, but that seems like the right compromise.